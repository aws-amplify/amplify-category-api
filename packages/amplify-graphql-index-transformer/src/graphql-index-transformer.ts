import {
  DirectiveWrapper,
  generateGetArgumentsInput,
  InvalidDirectiveError,
  TransformerPluginBase,
} from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerResolverProvider,
  TransformerSchemaVisitStepContextProvider,
  TransformerTransformSchemaStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { IndexDirective } from '@aws-amplify/graphql-directives';
import {
  DirectiveNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InterfaceTypeDefinitionNode,
  Kind,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { isListType, isScalarOrEnum } from 'graphql-transformer-common';
import { appendSecondaryIndex, constructSyncVTL, updateResolversForIndex } from './resolvers';
import { addKeyConditionInputs, ensureQueryField, updateMutationConditionInput } from './schema';
import { IndexDirectiveConfiguration } from './types';
import { generateKeyAndQueryNameForConfig, validateNotSelfReferencing } from './utils';

/**
 *
 */
export class IndexTransformer extends TransformerPluginBase {
  private directiveList: IndexDirectiveConfiguration[] = [];

  private resolverMap: Map<TransformerResolverProvider, string> = new Map();

  constructor() {
    super('amplify-index-transformer', IndexDirective.definition);
  }

  field = (
    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
    definition: FieldDefinitionNode,
    directive: DirectiveNode,
    context: TransformerSchemaVisitStepContextProvider,
  ): void => {
    const directiveWrapped = new DirectiveWrapper(directive);
    const args = directiveWrapped.getArguments(
      {
        object: parent as ObjectTypeDefinitionNode,
        field: definition,
        directive,
      } as IndexDirectiveConfiguration,
      generateGetArgumentsInput(context.transformParameters),
    );

    /**
     * Impute Optional Fields
     * Start with sort key fields since both the name and queryField will work sort key fields into the generated name if necessary.
     */
    args.sortKeyFields = getOrGenerateDefaultSortKeyFields(args);
    args.name = getOrGenerateDefaultName(args);
    args.queryField = getOrGenerateDefaultQueryField(context, args);
    args.sortKey = [];

    validate(args, context as TransformerContextProvider);
    this.directiveList.push(args);
  };

  public after = (ctx: TransformerContextProvider): void => {
    if (!ctx.isProjectUsingDataStore()) return;

    // construct sync VTL code
    this.resolverMap.forEach((syncVTLContent, resource) => {
      if (syncVTLContent) {
        constructSyncVTL(syncVTLContent, resource);
      }
    });
  };

  transformSchema = (ctx: TransformerTransformSchemaStepContextProvider): void => {
    const context = ctx as TransformerContextProvider;

    for (const config of this.directiveList) {
      ensureQueryField(config, context);
      addKeyConditionInputs(config, context);
      updateMutationConditionInput(config, context);
    }
  };

  generateResolvers = (ctx: TransformerContextProvider): void => {
    for (const config of this.directiveList) {
      appendSecondaryIndex(config, ctx);
      updateResolversForIndex(config, ctx, this.resolverMap);
    }
  };
}

/**
 * Return the name if provided in our args, else
 * compute the name based on the field name, and sortKeyFields.
 */
const getOrGenerateDefaultName = (config: IndexDirectiveConfiguration): string => {
  const indexNameRegex = /^[A-Za-z0-9_\-\.]{3,255}$/;
  if (config.name) {
    if (!indexNameRegex.test(config.name)) {
      throw new Error(
        `The indexName is invalid. It should be between 3 and 255 characters. Only A–Z, a–z, 0–9, underscore characters, hyphens, and periods allowed.`,
      );
    }
    return config.name;
  }

  if (config.name === null) {
    throw new Error('Explicit null value not allowed for name field on @index');
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * Return the queryField if provided in our args, else
 * compute the queryField based on the field name, and sortKeyFields if the feature flag is enabled.
 */
const getOrGenerateDefaultQueryField = (
  context: TransformerSchemaVisitStepContextProvider,
  config: IndexDirectiveConfiguration,
): string | null => {
  const autoIndexQueryNamesIsEnabled = context.transformParameters.enableAutoIndexQueryNames;
  // Any explicit null will take effect, if enableAutoIndexQueryNames and no queryField is provide set to null for consistency
  if (config.queryField === null || (!autoIndexQueryNamesIsEnabled && !config.queryField)) {
    return null;
  }

  if (config.queryField) {
    return config.queryField;
  }

  return generateKeyAndQueryNameForConfig(config);
};

/**
 * sortKeyFields are optional so default to empty list,
 * if we get a raw object just wrap in an array,
 * else return the list which was provided correctly.
 */
const getOrGenerateDefaultSortKeyFields = (config: IndexDirectiveConfiguration): string[] => {
  if (!config.sortKeyFields) {
    return [];
  }
  if (!Array.isArray(config.sortKeyFields)) {
    return [config.sortKeyFields];
  }
  return config.sortKeyFields;
};

const validate = (config: IndexDirectiveConfiguration, ctx: TransformerContextProvider): void => {
  const { name, object, field, sortKeyFields } = config;
  validateNotSelfReferencing(config);

  const modelDirective = object.directives!.find((directive) => directive.name.value === 'model');

  if (!modelDirective) {
    throw new InvalidDirectiveError(`The @${IndexDirective.name} directive may only be added to object definitions annotated with @model.`);
  }

  config.modelDirective = modelDirective;

  const fieldMap = new Map<string, FieldDefinitionNode>();

  for (const objectField of object.fields!) {
    fieldMap.set(objectField.name.value, objectField);

    for (const peerDirective of objectField.directives!) {
      if (peerDirective === config.directive) {
        continue;
      }

      if (peerDirective.name.value === 'primaryKey') {
        const hasSortFields = peerDirective.arguments!.some(
          (arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0,
        );
        config.primaryKeyField = objectField;

        if (!hasSortFields && objectField.name.value === field.name.value) {
          throw new InvalidDirectiveError(
            `Invalid @index '${name}'. You may not create an index where the partition key ` +
              'is the same as that of the primary key unless the primary key has a sort field. ' +
              'You cannot have a local secondary index without a sort key in the primary key.',
          );
        }
      }

      if (
        peerDirective.name.value === IndexDirective.name &&
        peerDirective.arguments!.some((arg: any) => arg.name.value === 'name' && arg.value.value === name)
      ) {
        throw new InvalidDirectiveError(
          `You may only supply one @${IndexDirective.name} with the name '${name}' on type '${object.name.value}'.`,
        );
      }
    }

    for (const peerDirective of objectField.directives!) {
      const hasSortFields = peerDirective.arguments!.some((arg: any) => arg.name.value === 'sortKeyFields' && arg.value.values?.length > 0);

      if (
        !ctx.transformParameters.secondaryKeyAsGSI &&
        !hasSortFields &&
        objectField == config.primaryKeyField &&
        objectField.name.value === field.name.value
      ) {
        throw new InvalidDirectiveError(
          `Invalid @index '${name}'. You may not create an index where the partition key ` +
            'is the same as that of the primary key unless the index has a sort field. ' +
            'You cannot have a local secondary index without a sort key in the index.',
        );
      }
    }
  }

  const enums = ctx.output.getTypeDefinitionsOfKind(Kind.ENUM_TYPE_DEFINITION) as EnumTypeDefinitionNode[];

  if (!isScalarOrEnum(field.type, enums) || isListType(field.type)) {
    throw new InvalidDirectiveError(`Index '${name}' on type '${object.name.value}.${field.name.value}' cannot be a non-scalar.`);
  }

  for (const sortKeyFieldName of sortKeyFields) {
    const sortField = fieldMap.get(sortKeyFieldName);

    if (!sortField) {
      throw new InvalidDirectiveError(
        `Can't find field '${sortKeyFieldName}' in ${object.name.value}, but it was specified in index '${name}'.`,
      );
    }

    if (!isScalarOrEnum(sortField.type, enums) || isListType(sortField.type)) {
      throw new InvalidDirectiveError(
        `The sort key of index '${name}' on type '${object.name.value}.${sortField.name.value}' cannot be a non-scalar.`,
      );
    }

    config.sortKey.push(sortField);
  }
};
