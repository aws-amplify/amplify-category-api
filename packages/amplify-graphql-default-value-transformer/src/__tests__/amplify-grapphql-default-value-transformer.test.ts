import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import {
  constructDataSourceStrategies,
  DDB_DEFAULT_DATASOURCE_STRATEGY,
  getResourceNamesForStrategy,
  MYSQL_DB_TYPE,
  POSTGRES_DB_TYPE,
  validateModelSchema,
} from '@aws-amplify/graphql-transformer-core';
import { parse } from 'graphql';
import { mockSqlDataSourceStrategy, testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { DefaultValueTransformer } from '..';

describe('DefaultValueModelTransformer:', () => {
  it('throws if @default is used in a non-@model type', () => {
    const schema = `
      type Test {
        id: ID!
        name: String @default(value: "hello world")
      }`;

    expect(() =>
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer()],
      }),
    ).toThrow('The @default directive may only be added to object definitions annotated with @model.');
  });

  it('throws if @default is used on a non scalar or enum field', () => {
    const schema = `
      type Test @model {
        id: ID!
        student: Student @default(value: "{'name':'FooBar'}")
      }

      type Student {
        name: String
      }
    `;

    expect(() =>
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer()],
      }),
    ).toThrow('The @default directive may only be added to scalar or enum field types.');
  });

  it.each([
    {
      type: 'String',
      value: undefined,
      expectedError: 'The @default directive requires a value property on non-Postgres datasources.',
    },
    { type: 'Int', value: '"text"', expectedError: 'Default value "text" is not a valid Int.' },
    { type: 'Boolean', value: '"text"', expectedError: 'Default value "text" is not a valid Boolean.' },
    { type: 'AWSJSON', value: '"text"', expectedError: 'Default value "text" is not a valid AWSJSON.' },
    { type: 'AWSDate', value: '"text"', expectedError: 'Default value "text" is not a valid AWSDate.' },
    { type: 'AWSDateTime', value: '"text"', expectedError: 'Default value "text" is not a valid AWSDateTime.' },
    { type: 'AWSTime', value: '"text"', expectedError: 'Default value "text" is not a valid AWSTime.' },
    { type: 'AWSTimestamp', value: '"text"', expectedError: 'Default value "text" is not a valid AWSTimestamp.' },
    { type: 'AWSURL', value: '"text"', expectedError: 'Default value "text" is not a valid AWSURL.' },
    { type: 'AWSPhone', value: '"text"', expectedError: 'Default value "text" is not a valid AWSPhone.' },
    { type: 'AWSIPAddress', value: '"text"', expectedError: 'Default value "text" is not a valid AWSIPAddress.' },
  ])(`throws if @default is used with invalid type. %type check.`, ({ type, value, expectedError }) => {
    const schema = `
      type Test @model {
      id: ID!
      value: ${type} ${value !== undefined ? `@default(value: ${value})` : '@default'}
      }
    `;
    expect(() =>
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer()],
      }),
    ).toThrow(expectedError);
  });

  it('should validate enum values', async () => {
    const inputSchema = `
      type Post @model {
        id: ID!
        enumValue: Tag @default(value: "INVALID")
      }

      enum Tag {
        NEWS
        RANDOM
      }
    `;

    expect(() =>
      testTransform({
        schema: inputSchema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer()],
      }),
    ).toThrow('Default value "INVALID" is not a member of Tag enum.');
  });

  it('should be supported on a required field.', () => {
    const inputSchema = `
      type Test @model {
        id: ID!
        stringValue: String! @default(value: "hello world")
      }
    `;

    const out = testTransform({
      schema: inputSchema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer()],
    });
    expect(out).toBeDefined();
    expect(out.schema).toMatchSnapshot();

    const schema = parse(out.schema);
    validateModelSchema(schema);
  });

  it('should successfully transform simple valid schema', async () => {
    const inputSchema = `
      type Post @model {
        id: ID!
        stringValue: String @default(value: "hello world")
        intVal: Int @default(value: "10002000")
        floatValue: Float @default(value: "123456.34565")
        booleanValue: Boolean @default(value: "true")
        awsJsonValue: AWSJSON @default(value: "{\\"a\\":1, \\"b\\":3, \\"string\\": 234}")
        awsDateValue: AWSDate @default(value: "2016-01-29")
        awsTimestampValue: AWSTimestamp @default(value: "545345345")
        awsEmailValue: AWSEmail @default(value: "local-part@domain-part")
        awsURLValue: AWSURL @default(value: "https://www.amazon.com/dp/B000NZW3KC/")
        awsPhoneValue: AWSPhone @default(value: "+41 44 668 18 00")
        awsIPAddressValue1: AWSIPAddress @default(value: "123.12.34.56")
        awsIPAddressValue2: AWSIPAddress @default(value: "1a2b:3c4b::1234:4567")
        enumValue: Tag @default(value: "RANDOM")
        awsTimeValue: AWSTime @default(value: "12:00:34Z")
        awsDateTime: AWSDateTime @default(value: "2007-04-05T14:30:34Z")
      }

      enum Tag {
        NEWS
        RANDOM
      }
    `;
    const out = testTransform({
      schema: inputSchema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer()],
    });
    expect(out).toBeDefined();
    expect(out.schema).toMatchSnapshot();

    const schema = parse(out.schema);
    validateModelSchema(schema);
  });

  it('should successfully set the default values when model name starts with lowercase', async () => {
    const inputSchema = `
      type post @model {
        id: ID!
        stringValue: String @default(value: "hello world")
      }
    `;
    const out = testTransform({
      schema: inputSchema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer()],
    });
    expect(out).toBeDefined();
    expect(out.resolvers).toBeDefined();
    expect(out.resolvers['Mutation.createPost.init.2.req.vtl']).toBeDefined();
    expect(out.resolvers['Mutation.createPost.init.2.req.vtl']).toMatchSnapshot();
    const schema = parse(out.schema);
    validateModelSchema(schema);
  });

  it('default value type should not be validated for sql datasource', async () => {
    const validSchema = `
      type Note @model {
          id: ID! @primaryKey
          content: String!
          createdAt: AWSDateTime @default(value: "CURRENT_TIMESTAMP")
      }
    `;

    const mySqlStrategy = mockSqlDataSourceStrategy();
    const resourceNames = getResourceNamesForStrategy(mySqlStrategy);
    const out = testTransform({
      schema: validSchema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(validSchema, mySqlStrategy),
    });
    expect(out).toBeDefined();

    validateModelSchema(parse(out.schema));
    expect(out.stacks).toBeDefined();
    expect(out.stacks[resourceNames.sqlStack]).toBeDefined();
    expect(out.stacks[resourceNames.sqlStack].Resources).toBeDefined();
    expect(out.resolvers['Mutation.createNote.init.1.req.vtl']).toBeDefined();
    expect(out.resolvers['Mutation.createNote.init.2.req.vtl']).toBeUndefined();
  });

  it('cannot use `null` as a default value', () => {
    const schema = `
      type CoffeeQueue @model {
        id: ID! @primaryKey
        orderNumber: Int! @default(value: null)
        name: String
      }`;

    expect(() => {
      testTransform({
        schema: schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
      });
    }).toThrow('The @default directive does not support null values.');
  });

  it.each([
    { strategy: mockSqlDataSourceStrategy({ dbType: MYSQL_DB_TYPE }), datasourceName: 'mysql' },
    { strategy: DDB_DEFAULT_DATASOURCE_STRATEGY, datasourceName: 'DynamoDB (Deafult)' },
  ])('throws if @default (no args) is used non-Postgres datasource (%datasourceName)', ({ strategy }) => {
    const schema = `
      type CoffeeQueue @model {
        id: ID! @primaryKey
        orderNumber: Int! @default
        name: String
      }`;
    expect(() => {
      testTransform({
        schema: schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
      });
    }).toThrow('The @default directive requires a value property on non-Postgres datasources.');
  });

  it.each([
    { typeStr: 'Boolean' },
    { typeStr: 'AWSJSON' },
    { typeStr: 'AWSDate' },
    { typeStr: 'AWSDateTime' },
    { typeStr: 'AWSTime' },
    { typeStr: 'AWSTime' },
    { typeStr: 'AWSURL' },
    { typeStr: 'AWSPhone' },
    { typeStr: 'AWSIPAddress' },
  ])('throws if @default (no args) is used on non-int types (case: $typeStr)', ({ typeStr }) => {
    const strategy = mockSqlDataSourceStrategy({ dbType: POSTGRES_DB_TYPE });

    expect(() => {
      const schema = `
      type Test @model {
        id: ID! @primaryKey
        value: ${typeStr} @default
      }
    `;
      testTransform({
        schema,
        transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
        dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
      });
    }).toThrow('The @default directive requires a value property on non-Int types.');
  });

  it('should successfully transform a simple schema with @default', async () => {
    const schema = `
      type TestAutoIncrement @model {
        id: ID! @primaryKey
        value: Int @default
      }
    `;
    const strategy = mockSqlDataSourceStrategy({ dbType: POSTGRES_DB_TYPE });
    const out = testTransform({
      schema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
    });
    expect(out).toBeDefined();
    expect(out.schema).toMatchSnapshot();
    const parsedSchema = parse(out.schema);
    validateModelSchema(parsedSchema);
  });

  it('should allow auto-increment primary key', async () => {
    const schema = `
      type CoffeeQueue @model {
        orderNumber: Int! @primaryKey @default
        order: String!
        customer: String
      }
    `;
    const strategy = mockSqlDataSourceStrategy({ dbType: POSTGRES_DB_TYPE });
    const out = testTransform({
      schema,
      transformers: [new ModelTransformer(), new DefaultValueTransformer(), new PrimaryKeyTransformer()],
      dataSourceStrategies: constructDataSourceStrategies(schema, strategy),
    });
    expect(out).toBeDefined();
    expect(out.schema).toMatchSnapshot();
    const parsedSchema = parse(out.schema);
    validateModelSchema(parsedSchema);
  });
});
