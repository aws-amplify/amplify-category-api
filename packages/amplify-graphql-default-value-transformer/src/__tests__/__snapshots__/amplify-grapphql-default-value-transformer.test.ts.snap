// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultValueModelTransformer: should be supported on a required field. 1`] = `
"type Test {
  id: ID!
  stringValue: String!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelTestConnection {
  items: [Test]!
  nextToken: String
}

input ModelTestFilterInput {
  id: ModelIDInput
  stringValue: ModelStringInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelTestFilterInput]
  or: [ModelTestFilterInput]
  not: ModelTestFilterInput
}

type Query {
  getTest(id: ID!): Test
  listTests(filter: ModelTestFilterInput, limit: Int, nextToken: String): ModelTestConnection
}

input ModelTestConditionInput {
  stringValue: ModelStringInput
  and: [ModelTestConditionInput]
  or: [ModelTestConditionInput]
  not: ModelTestConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreateTestInput {
  id: ID
  stringValue: String!
}

input UpdateTestInput {
  id: ID!
  stringValue: String
}

input DeleteTestInput {
  id: ID!
}

type Mutation {
  createTest(input: CreateTestInput!, condition: ModelTestConditionInput): Test
  updateTest(input: UpdateTestInput!, condition: ModelTestConditionInput): Test
  deleteTest(input: DeleteTestInput!, condition: ModelTestConditionInput): Test
}

input ModelSubscriptionTestFilterInput {
  id: ModelSubscriptionIDInput
  stringValue: ModelSubscriptionStringInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionTestFilterInput]
  or: [ModelSubscriptionTestFilterInput]
}

type Subscription {
  onCreateTest(filter: ModelSubscriptionTestFilterInput): Test @aws_subscribe(mutations: [\\"createTest\\"])
  onUpdateTest(filter: ModelSubscriptionTestFilterInput): Test @aws_subscribe(mutations: [\\"updateTest\\"])
  onDeleteTest(filter: ModelSubscriptionTestFilterInput): Test @aws_subscribe(mutations: [\\"deleteTest\\"])
}
"
`;

exports[`DefaultValueModelTransformer: should successfully set the default values when model name starts with lowercase 1`] = `
"## [Start] Setting \\"stringValue\\" to default value of \\"hello world\\". **
$util.qr($context.args.input.put(\\"stringValue\\", $util.defaultIfNull($ctx.args.input.stringValue, \\"hello world\\")))
## [End] Setting \\"stringValue\\" to default value of \\"hello world\\". **
{}"
`;

exports[`DefaultValueModelTransformer: should successfully transform simple valid schema 1`] = `
"type Post {
  id: ID!
  stringValue: String
  intVal: Int
  floatValue: Float
  booleanValue: Boolean
  awsJsonValue: AWSJSON
  awsDateValue: AWSDate
  awsTimestampValue: AWSTimestamp
  awsEmailValue: AWSEmail
  awsURLValue: AWSURL
  awsPhoneValue: AWSPhone
  awsIPAddressValue1: AWSIPAddress
  awsIPAddressValue2: AWSIPAddress
  enumValue: Tag
  awsTimeValue: AWSTime
  awsDateTime: AWSDateTime
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

enum Tag {
  NEWS
  RANDOM
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelPostConnection {
  items: [Post]!
  nextToken: String
}

input ModelTagInput {
  eq: Tag
  ne: Tag
}

input ModelPostFilterInput {
  id: ModelIDInput
  stringValue: ModelStringInput
  intVal: ModelIntInput
  floatValue: ModelFloatInput
  booleanValue: ModelBooleanInput
  awsJsonValue: ModelStringInput
  awsDateValue: ModelStringInput
  awsTimestampValue: ModelIntInput
  awsEmailValue: ModelStringInput
  awsURLValue: ModelStringInput
  awsPhoneValue: ModelStringInput
  awsIPAddressValue1: ModelStringInput
  awsIPAddressValue2: ModelStringInput
  enumValue: ModelTagInput
  awsTimeValue: ModelStringInput
  awsDateTime: ModelStringInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelPostFilterInput]
  or: [ModelPostFilterInput]
  not: ModelPostFilterInput
}

type Query {
  getPost(id: ID!): Post
  listPosts(filter: ModelPostFilterInput, limit: Int, nextToken: String): ModelPostConnection
}

input ModelPostConditionInput {
  stringValue: ModelStringInput
  intVal: ModelIntInput
  floatValue: ModelFloatInput
  booleanValue: ModelBooleanInput
  awsJsonValue: ModelStringInput
  awsDateValue: ModelStringInput
  awsTimestampValue: ModelIntInput
  awsEmailValue: ModelStringInput
  awsURLValue: ModelStringInput
  awsPhoneValue: ModelStringInput
  awsIPAddressValue1: ModelStringInput
  awsIPAddressValue2: ModelStringInput
  enumValue: ModelTagInput
  awsTimeValue: ModelStringInput
  awsDateTime: ModelStringInput
  and: [ModelPostConditionInput]
  or: [ModelPostConditionInput]
  not: ModelPostConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreatePostInput {
  id: ID
  stringValue: String
  intVal: Int
  floatValue: Float
  booleanValue: Boolean
  awsJsonValue: AWSJSON
  awsDateValue: AWSDate
  awsTimestampValue: AWSTimestamp
  awsEmailValue: AWSEmail
  awsURLValue: AWSURL
  awsPhoneValue: AWSPhone
  awsIPAddressValue1: AWSIPAddress
  awsIPAddressValue2: AWSIPAddress
  enumValue: Tag
  awsTimeValue: AWSTime
  awsDateTime: AWSDateTime
}

input UpdatePostInput {
  id: ID!
  stringValue: String
  intVal: Int
  floatValue: Float
  booleanValue: Boolean
  awsJsonValue: AWSJSON
  awsDateValue: AWSDate
  awsTimestampValue: AWSTimestamp
  awsEmailValue: AWSEmail
  awsURLValue: AWSURL
  awsPhoneValue: AWSPhone
  awsIPAddressValue1: AWSIPAddress
  awsIPAddressValue2: AWSIPAddress
  enumValue: Tag
  awsTimeValue: AWSTime
  awsDateTime: AWSDateTime
}

input DeletePostInput {
  id: ID!
}

type Mutation {
  createPost(input: CreatePostInput!, condition: ModelPostConditionInput): Post
  updatePost(input: UpdatePostInput!, condition: ModelPostConditionInput): Post
  deletePost(input: DeletePostInput!, condition: ModelPostConditionInput): Post
}

input ModelSubscriptionPostFilterInput {
  id: ModelSubscriptionIDInput
  stringValue: ModelSubscriptionStringInput
  intVal: ModelSubscriptionIntInput
  floatValue: ModelSubscriptionFloatInput
  booleanValue: ModelSubscriptionBooleanInput
  awsJsonValue: ModelSubscriptionStringInput
  awsDateValue: ModelSubscriptionStringInput
  awsTimestampValue: ModelSubscriptionIntInput
  awsEmailValue: ModelSubscriptionStringInput
  awsURLValue: ModelSubscriptionStringInput
  awsPhoneValue: ModelSubscriptionStringInput
  awsIPAddressValue1: ModelSubscriptionStringInput
  awsIPAddressValue2: ModelSubscriptionStringInput
  enumValue: ModelSubscriptionStringInput
  awsTimeValue: ModelSubscriptionStringInput
  awsDateTime: ModelSubscriptionStringInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionPostFilterInput]
  or: [ModelSubscriptionPostFilterInput]
}

type Subscription {
  onCreatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"createPost\\"])
  onUpdatePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"updatePost\\"])
  onDeletePost(filter: ModelSubscriptionPostFilterInput): Post @aws_subscribe(mutations: [\\"deletePost\\"])
}
"
`;
